moveit_ros/planning/planning_scene_monitor/test/trajectory_monitor_tests.cpp

```diff
@@ -0,0 +1,111 @@
+
+/* Author: Abishalini Sivaraman */
+
+#include <memory>
+#include <chrono>
+#include <thread>
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+#include "moveit/planning_scene_monitor/trajectory_monitor.h"
+#include "moveit/planning_scene_monitor/current_state_monitor.h"
+#include "moveit/utils/robot_model_test_utils.h"
+#include "rclcpp/node_interfaces/node_clock_interface.hpp"
+#include "rclcpp/node_interfaces/node_topics_interface.hpp"
+#include "rclcpp/rclcpp.hpp"
+#include "tf2_ros/buffer.h"
+
+using namespace std::chrono_literals;
+
+struct MockTrajectoryMonitorMiddlewareHandle : public planning_scene_monitor::TrajectoryMonitor::MiddlewareHandle
+{
+  MOCK_METHOD(void, sleep, (), (override));
+};
+
+struct MockCurrentStateMonitorMiddlewareHandle : public planning_scene_monitor::CurrentStateMonitor::MiddlewareHandle
+{
+  MOCK_METHOD(rclcpp::Time, now, (), (const, override));
+  MOCK_METHOD(void, createJointStateSubscription,
+              (const std::string& topic, planning_scene_monitor::JointStateUpdateCallback callback), (override));
+  MOCK_METHOD(void, resetJointStateSubscription, (), (override));
+  MOCK_METHOD(std::string, getJointStateTopicName, (), (const, override));
+};
+
+void waitFor(std::chrono::seconds timeout, std::function<bool()> done)
+{
+  const auto start = std::chrono::steady_clock::now();
+  while ((std::chrono::steady_clock::now() - start) < timeout)
+  {
+    if (done())
+    {
+      return;
+    }
+    std::this_thread::sleep_for(1ms);
+  }
+}
+
+TEST(TrajectoryMonitorTests, test1)
+{
+  auto mock_trajectory_monitor_middleware_handle = std::make_unique<MockTrajectoryMonitorMiddlewareHandle>();
+  auto mock_current_state_monitor_middleware_handle = std::make_unique<MockCurrentStateMonitorMiddlewareHandle>();
+
+  // THEN we expect it to call addSleep function
+  EXPECT_CALL(*mock_trajectory_monitor_middleware_handle, sleep).Times(::testing::AtLeast(1));
+
+  // GIVEN a TrajectoryMonitor is started
+  planning_scene_monitor::CurrentStateMonitorPtr current_state_monitor;
+
+  current_state_monitor = std::make_unique<planning_scene_monitor::CurrentStateMonitor>(
+      std::move(mock_current_state_monitor_middleware_handle), moveit::core::loadTestingRobotModel("panda"),
+      std::make_shared<tf2_ros::Buffer>(std::make_shared<rclcpp::Clock>()));
+
+  planning_scene_monitor::TrajectoryMonitor trajectory_monitor{ current_state_monitor,
+                                                                std::move(mock_trajectory_monitor_middleware_handle),
+                                                                10.0 };
+
+  std::atomic<bool> callback_called{ false };
+  trajectory_monitor.setOnStateAddCallback(
+      [&](const moveit::core::RobotStateConstPtr& robot_state, const rclcpp::Time& time) { callback_called = true; });
+
+  // WHEN we call the startTrajectoryMonitor function
+  trajectory_monitor.startTrajectoryMonitor();
+  waitFor(10s, [&]() { return callback_called == true; });
```

Comment:
Will this test run for 10 seconds? Probably should be in millisecond range? Or?

Reply:
It will wait for up to 10s for callback_called to equal true.  The actual time this takes is unknown but should be <1ms.